File: tools/commands/editors/neovim/run-real-server.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//!
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! ```

use std::{
    env, fs,
    path::PathBuf,
    process::{Command, ExitCode, Stdio},
};

#[path = "../../../rust/common.rs"]
mod common;
#[path = "../../../rust/editor_e2e.rs"]
mod editor_e2e;

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let repo = common::repo_root()?;
    let nvim = env::var("VIZE_TEST_NVIM_PATH").unwrap_or_else(|_| "nvim".to_string());
    let plugin_root = repo.join("editors/nvim");
    let spec_path = plugin_root.join("test/vize_e2e_spec.lua");
    let server_path = editor_e2e::resolve_real_server_path(&repo)?;
    let session = unique_temp_dir("vize-nvim-e2e")?;
    let workspace = session.join("real-vue");
    editor_e2e::prepare_real_vue_workspace(&workspace, false)?;
    let runtimepath = serde_json::to_string(&plugin_root.display().to_string())
        .map_err(|error| error.to_string())?;
    let spec = serde_json::to_string(&spec_path.display().to_string())
        .map_err(|error| error.to_string())?;
    let result = Command::new(nvim)
        .args([
            "--headless",
            "-u",
            "NONE",
            "--noplugin",
            "-n",
            "-i",
            "NONE",
            "-c",
            &format!("lua vim.opt.runtimepath:prepend({runtimepath})"),
            "-c",
            &format!("lua dofile({spec})"),
            "-c",
            "qall!",
        ])
        .current_dir(&workspace)
        .env("VIZE_E2E_SERVER", &server_path)
        .env("VIZE_E2E_WORKSPACE", &workspace)
        .stdin(Stdio::null())
        .output();
    let _ = fs::remove_dir_all(&session);
    let output = result.map_err(|error| format!("failed to run headless Neovim: {error}"))?;
    print!("{}", String::from_utf8_lossy(&output.stdout));
    eprint!("{}", String::from_utf8_lossy(&output.stderr));
    if !output.status.success() {
        return Err(format!(
            "headless Neovim scenario failed with exit code {}",
            output.status.code().unwrap_or(1)
        ));
    }
    Ok(())
}

fn unique_temp_dir(prefix: &str) -> Result<PathBuf, String> {
    let path = env::temp_dir().join(format!(
        "{}-{}-{}",
        prefix,
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map_err(|error| error.to_string())?
            .as_nanos()
    ));
    common::mkdir(&path)?;
    Ok(path)
}

File: tools/support/compat/nvim-vize/run-real-server.mjs
// Drive the packaged Neovim integration against a real `vize lsp` process
// (#3457). Neovim is the only non-VS Code editor in the packaged set with a
// scriptable headless LSP client, so it carries the second end-to-end scenario
// for the #3224 parity scorecard.
//
// This runner only prepares the workspace and launches headless Neovim; every
// assertion lives in `editors/nvim/test/vize_e2e_spec.lua`, which ships inside
// the Neovim tarball so a packaged checkout can run the same scenario.
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import {
  prepareRealVueWorkspace,
  repositoryRoot,
  resolveRealServerPath,
} from "../editor-e2e/real-vue-workspace.mjs";

const nvimPath = process.env.VIZE_TEST_NVIM_PATH?.trim() || "nvim";
const pluginRoot = path.join(repositoryRoot, "editors", "nvim");
const specPath = path.join(pluginRoot, "test", "vize_e2e_spec.lua");
const serverPath = resolveRealServerPath();
const sessionPath = fs.mkdtempSync(path.join(os.tmpdir(), "vize-nvim-e2e-"));
const workspacePath = path.join(sessionPath, "real-vue");

prepareRealVueWorkspace(workspacePath);

try {
  const result = spawnSync(
    nvimPath,
    [
      "--headless",
      "-u",
      "NONE",
      "--noplugin",
      "-n",
      "-i",
      "NONE",
      // Paths can contain spaces, backslashes or `|`, which ex commands would
      // mangle, so hand them to Lua as literals instead. The spec prepends the
      // plugin root itself, but do it here too so `require("vize.…")` works even
      // if the spec is ever loaded differently.
      "-c",
      `lua vim.opt.runtimepath:prepend(${JSON.stringify(pluginRoot)})`,
      "-c",
      `lua dofile(${JSON.stringify(specPath)})`,
      "-c",
      "qall!",
    ],
    {
      cwd: workspacePath,
      encoding: "utf-8",
      env: {
        ...process.env,
        VIZE_E2E_SERVER: serverPath,
        VIZE_E2E_WORKSPACE: workspacePath,
      },
      // The spec's own waits already add up to 960s in the worst case (120s
      // initialize + 240s diagnostics + 120s for each of the five requests and
      // the synchronous format), so this outer kill switch has to sit above
      // that or it would pre-empt the assertions it is meant to bound.
      timeout: 1_200_000,
    },
  );

  if (result.stdout) process.stdout.write(result.stdout);
  if (result.stderr) process.stderr.write(result.stderr);

  if (result.error) {
    throw result.error;
  }
  if (result.status !== 0) {
    throw new Error(`headless Neovim scenario failed with exit code ${result.status}`);
  }
} finally {
  fs.rmSync(sessionPath, { force: true, recursive: true });
}
